### PR TITLE
Add metadata support for TPC

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Part of #3170

- Added `UseTpcMappingStrategy` and `UseTptMappingStrategy`, calling either maps all derived to different tables/views
- Also just calling `ToTable` on a non-root type maps the hierarchy using TPT
- Throw for `ToTable` on abstract entity type if TPC
- If an entity type is not mapped to a table in TPT then the properties are mapped to the columns in the tables corresponding to its direct descendants and the identifying FK on the derived types is created pointing to the next mapped base type
- Only create check constraints on the first mapped type in TPT
- Added support for TPT and TPC for default mappings
- Warn that FKs that point to a TPC entity type won't be created in the database
- Warn when a PK property is store generated in TPC

### CODE REVIEW

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 1]
```diff
@@ -400,12 +406,18 @@
...
+    <value>The mapping strategy '{mappingStrategy}' specified on '{entityType}' is not supported.</value>
+  </data>
   <data name="InvalidMaxBatchSize" xml:space="preserve">
     <value>The specified 'MaxBatchSize' value '{value}' is not valid. It must be a positive number.</value>
   </data>
   <data name="InvalidMinBatchSize" xml:space="preserve">
     <value>The specified 'MinBatchSize' value '{value}' is not valid. It must be a positive number.</value>
   </data>
+  <data name="KeylessMappingStrategy" xml:space="preserve">
+    <value>The mapping strategy '{mappingStrategy}' used for '{entityType}' is not supported for keyless entity types.</value>
```
Comment (ajcvickers) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:419]: fwlink in this one?

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore.Relational/Extensions/RelationalEntityTypeExtensions.cs [hunk 2]
```diff
@@ -957,4 +965,58 @@ public static void SetIsTableExcludedFromMigrations(this IMutableEntityType enti
         this IConventionEntityType entityType)
         => entityType.FindAnnotation(RelationalAnnotationNames.IsTableExcludedFromMigrations)
             ?.GetConfigurationSource();
+
+    /// <summary>
+    ///     Gets the mapping strategy for the derived types.
+    /// </summary>
+    /// <param name="entityType">The entity type.</param>
+    /// <returns>The mapping strategy for the derived types.</returns>
+    public static string? GetMappingStrategy(this IReadOnlyEntityType entityType)
```
[the remark this answers is not in the document]
Comment (AndriySvyryd) [reply src/EFCore.Relational/Extensions/RelationalEntityTypeExtensions.cs:974]: Not yet. https://github.com/dotnet/efcore/issues/27642

File: src/EFCore.Relational/Diagnostics/RelationalLoggingDefinitions.cs [hunk 3]
```diff
@@ -484,6 +484,24 @@ public abstract class RelationalLoggingDefinitions : LoggingDefinitions
...
     public EventDefinitionBase? LogForeignKeyPropertiesMappedToUnrelatedTables;
 
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    [EntityFrameworkInternal]
+    public EventDefinitionBase? LogForeignKeyTPCPrincipal;
```
Comment (smitpatel) [inline src/EFCore.Relational/Diagnostics/RelationalLoggingDefinitions.cs:494]: Inconsistent TPC/Tpc in these 2.

**Comment by smitpatel** [comment]
>             root.Builder.HasNoDiscriminator();

What would happen in scenario when user configures discriminator and also applies TPT/TPC mapping strategy explicitly?

---
Refers to: src/EFCore.Relational/Metadata/Conventions/EntityTypeHierarchyMappingConvention.cs:110 in db75fae. [](commit_id = db75fae02aa6e2bf86ab9b31e18851371a6801e6, deletion_comment = False)

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs [hunk 4]
```diff
@@ -245,65 +245,80 @@ public override bool IsReadOnly
...
+        entityType.AddRuntimeAnnotation(RelationalAnnotationNames.DefaultMappings, tableMappings);
 
-        foreach (var property in entityType.GetProperties())
+        var isTpc = entityType.GetMappingStrategy() == RelationalAnnotationNames.TpcMappingStrategy;
+        var isTph = entityType.FindDiscriminatorProperty() != null;
+        while (mappedType != null)
         {
-            var columnName = property.GetColumnBaseName();
-            if (columnName == null)
+            var mappedTableName = isTph ? entityType.GetRootType().Name : mappedType.Name;
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:257]: Would have loved if the variable didn't use table in the name.

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs [hunk 5]
```diff
@@ -312,89 +327,98 @@ private static void AddDefaultMappings(RelationalModel databaseModel, IEntityTyp
...
 
-                if (mappedTableName == null
-                    || (mappedTableName == tableName
-                        && mappedSchema == schema
-                        && mappedType != entityType))
+        var mappedType = entityType;
+
+        Check.DebugAssert(entityType.FindRuntimeAnnotationValue(RelationalAnnotationNames.TableMappings) == null, "not null");
+        var tableMappings = new List<TableMapping>();
+        entityType.SetRuntimeAnnotation(RelationalAnnotationNames.TableMappings, tableMappings);
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:339]: With these now, all EntityTypes have table mapping regardless of them being populated. Is that intentional/required?

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs [hunk 6]
```diff
@@ -405,20 +429,28 @@ private static void AddViews(RelationalModel databaseModel, IEntityType entityTy
...
 
-        var schema = entityType.GetViewSchema();
-        List<ViewMapping>? viewMappings = null;
         var mappedType = entityType;
+
+        Check.DebugAssert(entityType.FindRuntimeAnnotationValue(RelationalAnnotationNames.ViewMappings) == null, "not null");
+        var viewMappings = new List<ViewMapping>();
+        entityType.SetRuntimeAnnotation(RelationalAnnotationNames.ViewMappings, viewMappings);
+
+        var isTpc = entityType.GetMappingStrategy() == RelationalAnnotationNames.TpcMappingStrategy;
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:438]: Can view be in TPC mapping?

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs [hunk 7]
```diff
@@ -1059,24 +1092,15 @@ private static void PopulateRowInternalForeignKeys(TableBase table)
...
-            mainMapping.IsSharedTablePrincipal = true;
-            ((View)mainMapping.Table).EntityTypeMappings.Add(mainViewMapping);
-        }
-
         Check.DebugAssert(
             mainMapping is not null,
             $"{nameof(mainMapping)} is neither a {nameof(TableMapping)} nor a {nameof(ViewMapping)}");
 
+        // Re-add the mapping to update the order
+        mainMapping.Table.EntityTypeMappings.Remove(mainMapping);
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:1100]: Not needed for view (as earlier)?

File: test/EFCore.Relational.Tests/Metadata/RelationalModelTest.cs [hunk 8]
```diff
@@ -23,54 +23,74 @@ public void GetRelationalModel_throws_if_convention_has_not_run()
...
         [InlineData(false, Mapping.TPT)]
+        [InlineData(false, Mapping.TPC)]
         public void Can_use_relational_model_with_tables(bool useExplicitMapping, Mapping mapping)
         {
             var model = CreateTestModel(mapToTables: useExplicitMapping, mapping: mapping);
 
-            Assert.Equal(9, model.Model.GetEntityTypes().Count());
-            Assert.Equal(mapping == Mapping.TPH || !useExplicitMapping ? 3 : 5, model.Tables.Count());
+            Assert.Equal(11, model.Model.GetEntityTypes().Count());
+            Assert.Equal(mapping == Mapping.TPC
```
Comment (smitpatel) [inline test/EFCore.Relational.Tests/Metadata/RelationalModelTest.cs:35]: could use switch expression.

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs (hunk 5 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:339]: Not required, but intentional. We could set them to the same reference if we find that this wastes memory.

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs (hunk 6 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:438]: Yes

**Comment by AndriySvyryd** [comment]
> What would happen in scenario when user configures discriminator and also applies TPT/TPC mapping strategy explicitly?

We throw `NonTphMappingStrategy`

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:257]: Yeah, me too...

**Pull request merged**